x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java

```diff
@@ -5,75 +5,323 @@
+            long from = lastRequestedSeqno + 1;
+            LOGGER.trace("{}[{}] peek read [{}]", params.getFollowShardId(), numConcurrentReads, from);
+            sendShardChangesRequest(from, maxReadSize, from);
+        }
+    }
+
+    private boolean hasReadBudget() {
+        assert Thread.holdsLock(this);
+        if (numConcurrentReads >= params.getMaxConcurrentReads()) {
+            LOGGER.trace("{} no new reads, maximum number of concurrent reads have been reached [{}]",
+                params.getFollowShardId(), numConcurrentReads);
+            return false;
+        }
+        if (buffer.size() > params.getMaxBufferSize()) {
+            LOGGER.trace("{} no new reads, buffer limit has been reached [{}]", params.getFollowShardId(), buffer.size());
+            return false;
+        }
+        return true;
+    }
+
+    private synchronized void coordinateWrites() {
+        while (hasWriteBudget() && buffer.isEmpty() == false) {
+            Translog.Operation[] ops = new Translog.Operation[Math.min(params.getMaxWriteSize(), buffer.size())];
+            for (int i = 0; i < ops.length; i++) {
+                ops[i] = buffer.remove();
+            }
+            numConcurrentWrites++;
+            LOGGER.trace("{}[{}] write [{}/{}] [{}]", params.getFollowShardId(), numConcurrentWrites, ops[0].seqNo(),
+                ops[ops.length - 1].seqNo(), ops.length);
+            sendBulkShardOperationsRequest(ops);
+        }
+    }
+
+    private boolean hasWriteBudget() {
+        assert Thread.holdsLock(this);
+        if (numConcurrentWrites >= params.getMaxConcurrentWrites()) {
+            LOGGER.trace("{} maximum number of concurrent writes have been reached [{}]",
+                params.getFollowShardId(), numConcurrentWrites);
+            return false;
+        }
+        return true;
+    }
+
+
+    private void sendShardChangesRequest(long from, int maxOperationCount, long maxRequiredSeqNo) {
+        innerSendShardChangesRequest(from, maxOperationCount,
+            response -> {
+                retryCounter.set(0);
+                handleReadResponse(from, maxOperationCount, maxRequiredSeqNo, response);
+            },
+            e -> handleFailure(e, () -> sendShardChangesRequest(from, maxOperationCount, maxRequiredSeqNo)));
+    }
+
+    private void handleReadResponse(long from, int maxOperationCount, long maxRequiredSeqNo, ShardChangesAction.Response response) {
+        maybeUpdateMapping(response.getIndexMetadataVersion(), () -> {
+            synchronized (ShardFollowNodeTask.this) {
+                globalCheckpoint = Math.max(globalCheckpoint, response.getLeaderCheckpoint());
+                buffer.addAll(Arrays.asList(response.getOperations()));
+                coordinateWrites();
+
+                OptionalLong lastOp = Arrays.stream(response.getOperations()).mapToLong(Translog.Operation::seqNo).max();
+                if (lastOp.isPresent() && lastOp.getAsLong() < maxRequiredSeqNo) {
```

Comment:
If shard changes api fails with an error we either fail or retry. In the retry case we never get in this method and only if it does succeed we get here.

 I think in the following situation we return 0 ops while we there are ops to be returned:
* MaxOperationSizeInBytes is set to a too low value and no ops can be returned.
* A leader shard copy's global checkpoint is equal to from.

I think the first case we should fail with an error in shard changes api if MaxOperationSizeInBytes is too low that we can't add a single op. We shouldn't retry this error.

The second case, I think is a scenario that can happen and by adding this check we would fail instead of sending an additional shard changes request. What do you think?

Reply:
Scratch the first case: https://github.com/elastic/elasticsearch/pull/31581#discussion_r200259758

Reply:
> The second case, I think is a scenario that can happen and by adding this check we would fail instead of sending an additional shard changes request. What do you think?

Sounds reasonable - please add a clear comment + a clear code block to deal with it and add it to the list of scenario's we want to test.

Reply:
I will add a comment, but I'm unsure how we can deal with it. How can we detect here that we got no ops,because we read from a leader shard copy with a global checkpoint equal to from; versus something went very bad and we should fail?